### ISSUE TITLE
Lint Rule for Unassigned Returned Values

### ISSUE THREAD

**Issue by ghost** [dart-lang/linter#2423]
Is there a lint rule for checking if there are unassigned returned values in the code? If not, is it possible for you guys to create it?

I and other developers have been working on a project for immutable collections called [fast_immutable_collections][fic] (FIC). Collections in that package do not mutate. If you wish to change anything, they will return a new object with the new data. So, for example:

```dart
IList<int> ilist = IList([1, 2, 3]); // an immutable list

IList<int> changedIlist = ilist.add(4);
```

That looks simple enough, but it's very easy for people used to the normal Dart API to forget to reassign the result of a change to a variable. If they don't, nothing happens:

```dart
IList<int> ilist = IList([1, 2, 3]); // an immutable list

// The person didn't realize this doesn't do anything to `ilist` itself, 
// the returned value will be garbage-collected right away:
ilist.add(4); 
```

So, is there a lint rule for this already? It doesn't look that esoteric...

[fic]: https://github.com/marcglasberg/fast_immutable_collections

**Comment by bwilkerson** [comment]
As far as I know, we don't have a lint for that. A general "always assign a returned value" lint would have a lot of false positives. For example, the method `Set.add` (from `dart:core`) returns a `bool` that's often not interesting to clients.

If we were going to support something like this then we'd likely want to add an annotation in `meta` that package authors like yourself could add to their APIs to indicate that the return value should never be ignored.

**Comment by Philpid** [comment]
It's true that there would be a lot of false positives, but I think it would be overall useful in spite of it because I don't believe they would normally undermine its utility, it would depend on what the programmer is working on. If he/she is working on something that interacts with a lot of false positives then he/she would always have the choice of leaving the rule turned off. 

At any rate, if I'm not mistaken, the programmer can always add a `// ignore: ...` to turn it off if there's a specific problem somewhere, can't he?

The `meta` annotation would be quite nice as well, and having both would provide much more flexibility indeed.

**Comment by srawlins** [comment]
Duplicate of #1888

**Comment by pq** [comment]
Let's consolidate this w/ the #1888.
